flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -2628,14 +2704,17 @@ class ScalarExprLowering {
     return gen(x);
   }
 
+  llvm::SmallVector<mlir::Attribute> &getRawVals() { return rawVals; }
 private:
   mlir::Location location;
   Fortran::lower::AbstractConverter &converter;
   fir::FirOpBuilder &builder;
   Fortran::lower::StatementContext &stmtCtx;
   Fortran::lower::SymMap &symMap;
   bool inInitializer;
-  bool useBoxArg = false; // expression lowered as argument
+  bool useBoxArg = false;                     // expression lowered as argument
+  llvm::SmallVector<mlir::Attribute> rawVals; // initialization raw values
+  int rawValWidth; // width of initialization raw values
```

Comment:
Have you consider making an independent function that takes Expr<SomeType>, unwraps a Constant<T> and generates the mlir::Attribute ?

If you make it return an llvm::Optional, you do not even need to have the users assume which types/rank are supported and could use a `if (auto attr = tryCreatingConstantAttribute(converter, expr)) createGlobal(..., *attr);` pattern instead.

In my opinion, avoiding overloading `ScalarExprLowering` might be best, normal expression lowering is not trivial to follow, adding new members won't help. I understand it is nice since it does the expression tree traversal for you though, so if what I am suggesting here does not make total sense, feel free to ignore it.

Reply:
I spent a lot of time trying to unwrap the constant with no success. Perhaps down the road with my more experience I can separate this into its own stand alone function.